jfr/src/main/java/io/netty/jfr/JfrChannelHandler.java

```diff
@@ -0,0 +1,383 @@
+                                   connectEvent.setLocalAddress(localAddress);
+                               })) {
+            connectEvent.commit();
+        }
+    }
+
+    @Override
+    public void disconnect(ChannelHandlerContext ctx, ChannelPromise promise) {
+        if (wrapInChannelEvent(disconnectEvent, ctx, () -> ctx.disconnect(promise))) {
+            disconnectEvent.commit();
+        }
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) {
+        if (wrapInChannelEvent(closeEvent, ctx, () -> ctx.close(promise))) {
+            closeEvent.commit();
+        }
+    }
+
+    @Override
+    public void deregister(ChannelHandlerContext ctx, ChannelPromise promise) {
+        if (wrapInChannelEvent(deregisterEvent, ctx, () -> ctx.deregister(promise))) {
+            deregisterEvent.commit();
+        }
+    }
+
+// NOTE: Occurs so often, currently removed for performance reasons, will try to optimize and add.
+//    @Override
+//    public void channelReadComplete(ChannelHandlerContext ctx) {
+//        if (wrapInChannelEvent(channelReadCompleteEvent, ctx, () ->  ctx::fireChannelReadComplete)) {
+//            channelReadCompleteEvent.commit();
+//        }
+//}
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        if (wrapInChannelEvent(channelReadEvent, ctx, () -> ctx.fireChannelRead(msg),
+                               () -> channelReadEvent.setMsg(msg), () -> {
+                })) {
+            channelReadEvent.commit();
+        }
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        if (wrapInChannelEvent(writeEvent, ctx, () -> ctx.write(msg, promise),
+                               () -> writeEvent.setMsg(msg), () -> {
+                })) {
+            writeEvent.commit();
+        }
+    }
+
+    @Override
+    public void channelWritabilityChanged(ChannelHandlerContext ctx) {
+        if (wrapInChannelEvent(channelWritabilityChangedEvent, ctx, ctx::fireChannelWritabilityChanged,
+                               () -> channelWritabilityChangedEvent.setWasWritable(ctx.channel().isWritable()),
+                               () -> channelWritabilityChangedEvent.setIsWritable(ctx.channel().isWritable()))) {
+            channelWritabilityChangedEvent.commit();
+        }
+    }
+
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+        if (wrapInChannelEvent(flushEvent, ctx, ctx::flush)) {
+            flushEvent.commit();
+        }
+    }
+
+    // Helper methods
+
+    private boolean wrapInChannelEvent(ChannelEvent event, ChannelHandlerContext ctx, Runnable operation) {
+        return wrapInChannelEvent(event, ctx, operation, () -> {
+        }, () -> {
+        });
+    }
+
+    private boolean wrapInChannelEvent(ChannelEvent event, ChannelHandlerContext ctx, Runnable operation,
```

Comment:
static

Reply:
Sorry for you having to do all these nitpicks..